### PR TITLE
Reimplement cc_attention using einops.

### PR DESCRIPTION
**Pull request by Leojc**
## Motivation
Reimplement a faster and more readable cc_attention using einops. 

## Modification
1. Rewrite `cc_attention.py`
2. Add einops to the `requirement.txt`

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Hi @Leojc ，
1. Do you have some statistic numbers that verifies the current implementation?
2. I understand that the CUDA ops might be replaced, but the interfaces of this module should not be changed, e.g., in_channels should not be changed to in_dims.

**Comment by Leojc** [comment]
Hi @ZwwWayne ,
1. 
    (1) I calculate the time for one forward (input size is [2,512,97,97]). The CUDA version is 0.0299619 and my implementation is 0.00544943. So about 5x faster.
    (2) I also test it in https://github.com/open-mmlab/mmsegmentation. I replace the cc_attention of ccnet in mmsegmentation. Then I run with a config of `ccnet_r50-d8_769x769_40k_cityscapes.py`. The CUDA version gets 78.46 mIoU, and this implementation gets 78.99. 
2. Sorry about that! I just fixed it.

**Comment by ZwwWayne** [comment]
Thank you @Leojc , the statistical results are important. You may also highlight the speed comparison in the docstring of the module. I will invite @xvjiarui to have look.

File: mmcv/ops/cc_attention.py [hunk 1]
```diff
@@ -1,91 +1,79 @@
...
+    The diagonal are all "-inf". This is for avoiding calculating the
+    overlapped element in the Criss-Cross twice.
+    """
+    return torch.diag(torch.tensor(float('-inf')).to(device).repeat(n), 0)
 
 
 class CrissCrossAttention(nn.Module):
-    """Criss-Cross Attention Module."""
+    """Criss-Cross Attention Module.
+
```
Comment (ZwwWayne) [inline mmcv/ops/cc_attention.py:18]: Suggest to also update arguments in the docstring.

Comment (Leojc) [reply mmcv/ops/cc_attention.py:18]: OK.

**Comment by hellock** [comment]
Can it be implemented with `torch.einsum` so that we can omit a third-party dependency?

**Comment by Leojc** [comment]
> 
> 
> Can it be implemented with `torch.einsum` so that we can omit a third-party dependency?

Yes, but that will decrease the readability greatly. Is that OK?

**Comment by Leojc** [comment]
> 
> 
> > Can it be implemented with `torch.einsum` so that we can omit a third-party dependency?
> 
> Yes, but that will decrease the readability greatly. Is that OK?

Wait... Maybe that won't decrease too much. I'll try implementing it.

**Comment by Leojc** [comment]
Think I'll open a new PR.
Reimplement cc_attention using pure pytorch #1201

**Comment by hellock** [comment]
> Think I should open a new PR.
> Reimplement cc_attention using pure PyTorch #1201

Looks great!

**Pull request closed without merging**